sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/LogicalQueryStageStrategy.scala

```diff
@@ -50,9 +52,34 @@ object LogicalQueryStageStrategy extends Strategy with PredicateHelper {
 
     case j @ Join(left, right, joinType, condition, _)
         if isBroadcastStage(left) || isBroadcastStage(right) =>
-      val buildSide = if (isBroadcastStage(left)) BuildLeft else BuildRight
-      BroadcastNestedLoopJoinExec(
-        planLater(left), planLater(right), buildSide, joinType, condition) :: Nil
+      def createBroadcastNLJoin() = {
+        val buildSide = if (isBroadcastStage(left)) BuildLeft else BuildRight
+        BroadcastNestedLoopJoinExec(
+          planLater(left), planLater(right), buildSide, joinType, condition) :: Nil
+      }
+
+      /**
+       * See. [SPARK-32290]
+       * Not in Subquery will almost certainly be planned as a Broadcast Nested Loop join,
+       * which is very time consuming because it's an O(M*N) calculation.
+       * But if it's a single column NotInSubquery, and buildSide data is small enough,
+       * O(M*N) calculation could be optimized into O(M) using hash lookup instead of loop lookup.
+       */
+      if (SQLConf.get.notInSubqueryHashJoinEnabled &&
```

Comment:
I am not sure if this is possible or what the prior convention here is: But would it be possible to share this section of the code with SparkStrategies.scala ? The only thing difference is the fallback to createBroadcastNLJoin() and I think that may be able to factored out.

Reply:
updated to new pattern match code.